nifi-extension-bundles/nifi-extension-utils/nifi-git-flow-registry/src/main/java/org/apache/nifi/registry/flow/git/AbstractGitFlowRegistryClient.java

```diff
@@ -0,0 +1,644 @@
+            final RegisteredFlow registeredFlow = new RegisteredFlow();
+            registeredFlow.setName(flowLocation.getFlowId());
+            registeredFlow.setIdentifier(flowLocation.getFlowId());
+            flowSnapshot.setFlow(registeredFlow);
+        }
+        if (flowSnapshot.getSnapshotMetadata() == null) {
+            final RegisteredFlowSnapshotMetadata snapshotMetadata = new RegisteredFlowSnapshotMetadata();
+            snapshotMetadata.setFlowIdentifier(flowLocation.getFlowId());
+            flowSnapshot.setSnapshotMetadata(snapshotMetadata);
+        }
+    }
+
+    private void verifyWritePermissions(final GitRepositoryClient repositoryClient) throws AuthorizationException {
+        if (!repositoryClient.getCanWrite()) {
+            throw new AuthorizationException("Client does not have write access to the repository");
+        }
+    }
+
+    private void verifyReadPermissions(final GitRepositoryClient repositoryClient) throws AuthorizationException {
+        if (!repositoryClient.getCanRead()) {
+            throw new AuthorizationException("Client does not have read access to the repository");
+        }
+    }
+
+    protected synchronized GitRepositoryClient getRepositoryClient(final FlowRegistryClientConfigurationContext context) throws IOException, FlowRegistryException {
+        if (!clientInitialized.get()) {
+            getLogger().info("Initializing repository client");
+            repositoryClient = createRepositoryClient(context);
+            clientInitialized.set(true);
+            initializeDefaultBucket(context);
+        }
+        return repositoryClient;
+    }
+
+    protected void invalidateClient() {
+        clientInitialized.set(false);
+        if (repositoryClient != null) {
+            try {
+                repositoryClient.close();
+            } catch (final Exception e) {
+                getLogger().warn("Error closing repository client", e);
+            }
+        }
+        repositoryClient = null;
+    }
+
+    // If the client has write permissions to the repo, then ensure the directory for the default bucket is present and if not create it,
+    // otherwise the client can only be used to import flows from the repo and won't be able to set up the default bucket
+    private void initializeDefaultBucket(final FlowRegistryClientConfigurationContext context) throws IOException, FlowRegistryException {
+        if (!repositoryClient.getCanWrite()) {
+            getLogger().info("Repository client does not have write permissions to the repository, skipping setup of default bucket");
+            return;
+        }
+
+        final String branch = context.getProperty(REPOSITORY_BRANCH).getValue();
+        final Set<String> bucketDirectoryNames = repositoryClient.getTopLevelDirectoryNames(branch);
+        if (!bucketDirectoryNames.isEmpty()) {
+            getLogger().info("Found existing buckets, skipping setup of default bucket");
```

Comment:
Is this necessary to be an `info` message, or should it be a `debug`? Does it make sense to include the bucketDirectoryNames? Perhaps just the number of directories would be helpful.